### PR TITLE
feat: improve support for `pytorch_geometric` and custom pytorch batches

### PR DESCRIPTION
**Pull request by ioga**
## Description

There was some interest in OSS slack on `pytorch_geometric` support; the problem is that this library uses custom class which exposes `.to` method (so we can move it to the device just fine), but its `__len__` doesn't represent the number of records in a batch, and we require that value.

Let's allow users to override the batch length calculation in using `PyTorchTrial#calculate_batch_length`.

## Test Plan

This PR includes an example based on `pytorch_geometric`, which is also our first GCN (graph convolutional network) example. It should run to completion on a cluster both in `const` and `distributed` configurations. The standard trials which do not override this field should run normally as well.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by shiyuann** [comment]
TBH, I think `calculate_ batch_length` is a hack to override the behavior of the `TrialController`. And it's hard to understand as an user interface. We should actually expose `TrialController` to users rather than letting them need to understand a function that is not useful in most cases.

**Comment by ioga** [comment]
>need to understand a function that is not useful in most cases
I don't think they need to understand this function for most cases, just for the ones where they actually need it, like this one. In that regard, it's similar to other more advanced features beyond the bare minimum, such as callbacks or `evaluation_reducer`.

>I think our API designing framework of letting the Trial class be only user-facing and TrialController to be internal is wrong. Any hacks to make up the flaws of this framework are making the API worse.
So what is your proposal? 

a. Move `calculate_batch_length` to `PyTorchTrialController` , so a user would do something like

```
class MyTrialController(PyTorchTrialController):
    def calculate_batch_length(self, batch):
        return batch.num_graphs

class MyTrial(PyTorchTrial):
    trial_controller_class = MyTrialController
```

b. not expose this as abstraction at all, and make them override entire `_train_for_step` and `_compute_validation_metrics`?

File: examples/graphs/proteins_pytorch_geometric/model_def.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        # other when doing distributed training.
+        download_directory = tempfile.mkdtemp()
+
+        self.dataset = TUDataset(
+            root=download_directory,
+            name=self.context.get_hparam("dataset"),
+            use_node_attr=True,
+        )
+
+        val_ratio = 0.2
```
Comment (naren-determined) [inline examples/graphs/proteins_pytorch_geometric/model_def.py:69]: Can we put this in the .yaml file config? I don't like hardcoded values like this in the model definition.

**Review by naren-determined — COMMENTED** [review]
Similar to Ilia's stance, I view `calculate_batch_length` as an advanced feature that is optionally used just like custom reducers. I don't really think exposing the `TrialController` is going to be that useful for the vast majority of users who may not necessarily be using these advanced features and may in fact cause more confusion. Not a fan of options (a) or (b) Ilia laid out.

File: examples/graphs/proteins_pytorch_geometric/distributed.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  hidden_channels: 32
+records_per_epoch: 890
+searcher:
+  name: single
+  metric: validation_loss
+  max_length:
+    epochs: 300
+  smaller_is_better: true
+entrypoint: model_def:GCNTrial
+resources:
```
Comment (shiyuann) [inline examples/graphs/proteins_pytorch_geometric/distributed.yaml:16]: remove this?

File: examples/graphs/proteins_pytorch_geometric/distributed.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  slots_per_trial: 4
+environment:
```
Comment (shiyuann) [inline examples/graphs/proteins_pytorch_geometric/distributed.yaml:18]: also remove this?

File: harness/determined/pytorch/_pytorch_trial.py [hunk 4]
```diff
@@ -980,6 +980,36 @@ def evaluate_full_dataset(self, data_loader: torch.utils.data.DataLoader) -> Dic
         """
         pass
 
+    def calculate_batch_length(self, batch: Any) -> int:
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_trial.py:983]: rename it to use "get" since it doesn't really calculate anything?

```suggestion
    def get_batch_length(self, batch: Any) -> int:
```

File: examples/graphs/proteins_pytorch_geometric/const.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  hidden_channels: 32
+records_per_epoch: 1001
+searcher:
+  name: single
+  metric: validation_loss
+  max_length:
+    epochs: 300
+  smaller_is_better: true
+entrypoint: model_def:GCNTrial
+environment:
```
Comment (shiyuann) [inline examples/graphs/proteins_pytorch_geometric/const.yaml:16]: same

**Review by shiyuann — APPROVED** [review]
LGTM!
